### PR TITLE
fix: Improve Timers error handling and web compatibility (fixes #45085)

### LINKED ISSUE #45085: Timers in the new architecture are not spec compliant
### Description

`setTimeout`, `clearTimeout`, `setInterval` and `clearInterval` all have a different behavior regarding error handling compared to Web.

On Web, `setTimeout` and `clearTimeout` only throw when the callback argument isn't provided. They don't even throw when the callback isn't a function. `clearTimeout` and `clearInterval` never throw.

RN (in the new architecture) throws in many more cases:
* `setTimeout` and `setInterval` throw is any of the arguments is missing, or has the wrong type.
* `clearTimeout` and `clearInterval` throw if the argument is missing, isn't a number or isn't a registered timer.

We should align these behaviors with the Web (more details in https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#dom-settimeout-dev).

### React Native Version

0.74.2 / main

### Affected Platforms

Runtime - Android, Runtime - iOS, Runtime - Web, Runtime - Desktop

### Areas

Bridgeless - The New Initialization Flow, Other (please specify)

### PR DESCRIPTION
**Pull request by robik**
## Summary:

Improve compatibility with web implementations of JS timers.

Fixes #45085

## Changelog:

[GENERAL] [CHANGED] - Timer functions are now throwing exceptions in less cases and are instead quiet quitting (similar to browsers)
[GENERAL] [CHANGED] - Timer functions `timeout` argument is now coerced to a number

## Test Plan:

Updated RN tester

### CODE REVIEW

File: packages/react-native/ReactCommon/react/runtime/TimerManager.cpp [hunk 1]
```diff
@@ -12,6 +12,26 @@
...
+  
+  // fast-path
+  if (timeout.isNumber()) {
+    delay = timeout.getNumber();
+  }
+  else {
+    // perform number coercion for timeout to be web spec compliant
+    auto numberCtor = rt.global().getPropertyAsObject(rt, "Number");
+    auto delayNumberObject = numberCtor.getFunction(rt).callAsConstructor(rt, timeout, 1).getObject(rt);
+    auto delayNumericValue = delayNumberObject.getPropertyAsFunction(rt, "valueOf").callWithThis(rt, delayNumberObject, nullptr, 0);
```
Comment (javache) [inline packages/react-native/ReactCommon/react/runtime/TimerManager.cpp:26]:
```suggestion
    auto delayNumericValue = delayNumberObject.getPropertyAsFunction(rt, "valueOf").callWithThis(rt, delayNumberObject);
```

**Review by javache — CHANGES_REQUESTED** [review]
Thanks for improving this, a few small tweaks please.

File: packages/react-native/ReactCommon/react/runtime/TimerManager.cpp [hunk 2]
```diff
@@ -12,6 +12,26 @@
 
 namespace facebook::react {
 
+double coerceNumberTimeout(jsi::Runtime& rt, jsi::Value&& timeout) {
```
[the remark this answers is not in the document]
Comment (robik) [reply packages/react-native/ReactCommon/react/runtime/TimerManager.cpp]: Oh, right. This is a leftover, forgot to change this. Good catch!

File: packages/react-native/ReactCommon/react/runtime/TimerManager.cpp [hunk 3]
```diff
@@ -344,8 +357,7 @@ void TimerManager::attachGlobals(jsi::Runtime& runtime) {
                   rt, "The first argument to setInterval must be a function.");
             }
             auto callback = args[0].getObject(rt).getFunction(rt);
-            auto delay =
-                count > 1 && args[1].isNumber() ? args[1].getNumber() : 0;
+            auto delay = count > 1 ? coerceNumberTimeout(rt, jsi::Value { rt, args[1] }) : 0.0;
```
Comment (javache) [inline packages/react-native/ReactCommon/react/runtime/TimerManager.cpp:360]: There should be no need to create a new jsi::Value here
```suggestion
            auto delay = count > 1 ? coerceNumberTimeout(rt, args[1]) : 0.0;
```

**Review by javache — APPROVED** [review]

**Pull request closed without merging**